File: lib/indexing/versions.ts
import versions from "@/lib/system-versions.json";
import type { ComponentVersionRow, TreeRow } from "@/lib/documents/detail";
import type { DocumentRow, DocumentStatus, IndexingRunRow } from "@/lib/documents";

export const SYSTEM_COMPONENT_VERSIONS = versions;

export type SystemComponent = keyof typeof SYSTEM_COMPONENT_VERSIONS;

export const SYSTEM_COMPONENT_VERSION_ROWS = Object.entries(SYSTEM_COMPONENT_VERSIONS).map(
  ([component, version]) => ({
    component,
    version
  })
);

export const TREE_INDEXER_PYTHON_ID = "sda-pageindex-python-langgraph";

export const TREE_INDEXER_PYTHON_VERSION =
  `${TREE_INDEXER_PYTHON_ID}-v${SYSTEM_COMPONENT_VERSIONS.tree_indexer_python}`;

export const INDEXING_VERSION_COLUMNS = {
  embedding_pipeline_version: SYSTEM_COMPONENT_VERSIONS.embedding_pipeline,
  extraction_pipeline_version: SYSTEM_COMPONENT_VERSIONS.extraction_pipeline,
  indexing_pipeline_version: SYSTEM_COMPONENT_VERSIONS.indexing_pipeline,
  tree_indexer_version: SYSTEM_COMPONENT_VERSIONS.tree_indexer_python
} as const;

export const INDEXING_VERSION_METADATA = {
  versions: {
    ...INDEXING_VERSION_COLUMNS,
    app_version: SYSTEM_COMPONENT_VERSIONS.app,
    compute_gateway_extraction_version:
      SYSTEM_COMPONENT_VERSIONS.compute_gateway_extraction,
    inngest_indexing_workflow_version:
      SYSTEM_COMPONENT_VERSIONS.inngest_indexing_workflow,
    tree_indexer_runtime_version: TREE_INDEXER_PYTHON_VERSION,
    tree_prompt_version: SYSTEM_COMPONENT_VERSIONS.tree_prompt
  }
} as const;

export type PipelineVersionKey =
  | "embedding_pipeline"
  | "extraction_pipeline"
  | "indexing_pipeline"
  | "tree_indexer"
  | "tree_prompt";

export type PipelineVersions = Record<PipelineVersionKey, string | null>;

export const PIPELINE_VERSION_COMPONENTS = [
  ["indexing_pipeline", "indexing_pipeline"],
  ["extraction_pipeline", "extraction_pipeline"],
  ["tree_indexer_python", "tree_indexer"],
  ["tree_prompt", "tree_prompt"],
  ["embedding_pipeline", "embedding_pipeline"]
] as const satisfies ReadonlyArray<readonly [string, PipelineVersionKey]>;

export function latestVersionMap(componentVersions: ComponentVersionRow[]) {
  return new Map(componentVersions.map((row) => [row.component, row.version]));
}

export function documentPipelineVersions(input: {
  document: DocumentRow;
  latestRun: IndexingRunRow | null;
  tree: TreeRow | null;
}): PipelineVersions {
  return {
    embedding_pipeline:
      input.document.embedding_pipeline_version ?? input.latestRun?.embedding_pipeline_version ?? null,
    extraction_pipeline:
      input.document.extraction_pipeline_version ?? input.latestRun?.extraction_pipeline_version ?? null,
    indexing_pipeline:
      input.document.indexing_pipeline_version ??
      input.tree?.indexing_pipeline_version ??
      input.latestRun?.indexing_pipeline_version ??
      null,
    tree_indexer:
      input.document.tree_indexer_version ??
      input.tree?.tree_indexer_version ??
      input.latestRun?.tree_indexer_version ??
      null,
    tree_prompt: input.tree?.tree_prompt_version ?? null
  };
}

export function isPipelineVersionStale(input: {
  documentStatus: DocumentStatus;
  latestVersions: Map<string, string>;
  versions: PipelineVersions;
}) {
  if (input.documentStatus !== "indexed") {
    return false;
  }

  return PIPELINE_VERSION_COMPONENTS.some(([component, versionKey]) => {
    const current = input.versions[versionKey];
    const latest = input.latestVersions.get(component);

    return Boolean(current && latest && current !== latest);
  });
}

export function pipelineVersionState(
  latestVersions: Map<string, string>,
  component: string,
  current: string | null
) {
  const latest = latestVersions.get(component);

  if (!current || !latest) {
    return "missing" as const;
  }

  return current === latest ? ("current" as const) : ("stale" as const);
}

File: workers/tree-indexer-python/app/versions.py
from __future__ import annotations

import json
import os
from pathlib import Path


def _load_versions() -> dict[str, str]:
    candidates = [
        os.getenv("SDA_SYSTEM_VERSIONS_FILE"),
        Path(__file__).resolve().parents[3] / "lib" / "system-versions.json",
        Path(__file__).resolve().parents[1] / "system-versions.json",
    ]

    for candidate in candidates:
        if not candidate:
            continue

        try:
            with open(candidate, encoding="utf-8") as source:
                data = json.load(source)
        except OSError:
            continue

        return {key: value for key, value in data.items() if isinstance(value, str)}

    return {}


def _version(name: str, fallback: str) -> str:
    return os.getenv(name, fallback)


_VERSIONS = _load_versions()

SYSTEM_COMPONENT_VERSIONS = {
    **_VERSIONS,
    "tree_indexer_python": _version(
        "SDA_TREE_INDEXER_VERSION",
        _VERSIONS.get("tree_indexer_python", "0.0.0"),
    ),
    "tree_prompt": _version(
        "SDA_TREE_PROMPT_VERSION",
        _VERSIONS.get("tree_prompt", "0.0.0"),
    ),
}

TREE_INDEXER_PYTHON_ID = "sda-pageindex-python-langgraph"
TREE_INDEXER_PYTHON_VERSION = (
    f"{TREE_INDEXER_PYTHON_ID}-v{SYSTEM_COMPONENT_VERSIONS['tree_indexer_python']}"
)

INDEXING_VERSION_COLUMNS = {
    "embedding_pipeline_version": SYSTEM_COMPONENT_VERSIONS.get(
        "embedding_pipeline",
        "0.0.0",
    ),
    "extraction_pipeline_version": SYSTEM_COMPONENT_VERSIONS.get(
        "extraction_pipeline",
        "0.0.0",
    ),
    "indexing_pipeline_version": SYSTEM_COMPONENT_VERSIONS.get(
        "indexing_pipeline",
        "0.0.0",
    ),
    "tree_indexer_version": SYSTEM_COMPONENT_VERSIONS["tree_indexer_python"],
}

TREE_PROMPT_VERSION = SYSTEM_COMPONENT_VERSIONS["tree_prompt"]


def version_value(versions: dict[str, str] | None, key: str, fallback: str) -> str:
    value = (versions or {}).get(key)
    return value if isinstance(value, str) and value else fallback
